File: Android Project/app/src/main/java/com/example/histoquiz/activities/EnterLocalGameActivity.java
package com.example.histoquiz.activities;

import androidx.appcompat.app.AppCompatActivity;
import com.example.histoquiz.R;
import com.google.android.material.textfield.TextInputLayout;
import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.firestore.DocumentSnapshot;
import com.google.firebase.firestore.FieldPath;
import com.google.firebase.firestore.FirebaseFirestore;

import android.content.Intent;
import android.os.Bundle;
import android.util.Log;
import android.widget.Button;
import android.widget.Toast;

import java.util.ArrayList;
import java.util.Objects;

public class EnterLocalGameActivity extends AppCompatActivity {

    public FirebaseFirestore firestoreDatabase;
    protected Button joinRoom;
    protected TextInputLayout roomCode;


    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_enter_local_game);
        initGUI();
    }

    protected void initGUI(){
        firestoreDatabase = FirebaseFirestore.getInstance();
        joinRoom = findViewById(R.id.entrarNaSala);
        roomCode = findViewById(R.id.codSala);
        joinRoom.setOnClickListener(v -> {
            if(Objects.requireNonNull(roomCode.getEditText()).getText().toString().isEmpty()){
                Toast.makeText(EnterLocalGameActivity.this, "Informe um código de sala.", Toast.LENGTH_LONG).show();
            }
            else{
                joinGameRoom(roomCode.getEditText().getText().toString());
            }
        });
    }

    @SuppressWarnings("unchecked")
    protected void joinGameRoom(String roomCode){
        firestoreDatabase.collection("partidaLocal").whereEqualTo(FieldPath.documentId(), roomCode).get().addOnCompleteListener(task -> {
            if (task.isSuccessful()) {
                if(!Objects.requireNonNull(task.getResult()).getDocuments().isEmpty()) {
                    DocumentSnapshot document = Objects.requireNonNull(task.getResult()).getDocuments().get(0);
                    if (document.exists()) {
                        int qntdPlayers = Integer.parseInt(Objects.requireNonNull(document.get("qntd")).toString());
                        if (qntdPlayers >= 4) {
                            Toast.makeText(EnterLocalGameActivity.this, "Essa sala já está cheia!", Toast.LENGTH_LONG).show();
                        } else {
                            ArrayList<String> nomeJogadores = (ArrayList<String>) document.get("nomeJogadores");
                            assert nomeJogadores != null;
                            nomeJogadores.set(qntdPlayers, Objects.requireNonNull(FirebaseAuth.getInstance().getCurrentUser()).getDisplayName());
                            firestoreDatabase.collection("partidaLocal").document(document.getId()).update("nomeJogadores", nomeJogadores);

                            ArrayList<String> uidJogadores = (ArrayList<String>) document.get("uidJogadores");
                            assert uidJogadores != null;
                            uidJogadores.set(qntdPlayers, FirebaseAuth.getInstance().getCurrentUser().getUid());
                            firestoreDatabase.collection("partidaLocal").document(document.getId()).update("uidJogadores", uidJogadores);

                            firestoreDatabase.collection("partidaLocal").document(document.getId()).update("qntd", Integer.toString(Integer.parseInt(Objects.requireNonNull(document.get("qntd")).toString()) + 1));

                            Intent troca = new Intent(EnterLocalGameActivity.this, LocalGameActivity.class);
                            troca.putExtra("matchCreator", false);
                            troca.putExtra("roomCode", roomCode);
                            startActivityForResult(troca, 999);
                        }
                    }
                }
                else {
                    Toast.makeText(EnterLocalGameActivity.this, "Essa sala não existe!", Toast.LENGTH_LONG).show();
                }
            } else {
                Log.d("Erro ao entrar na sala", "Erro: ", task.getException());
            }
        });
    }
}

File: Android Project/app/src/main/java/com/example/histoquiz/activities/EnterLocalGameActivity.kt
package com.example.histoquiz.activities

import android.content.Intent
import android.os.Bundle
import android.util.Log
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import androidx.core.view.WindowCompat
import androidx.core.view.WindowInsetsCompat
import androidx.core.view.WindowInsetsControllerCompat
import com.example.histoquiz.databinding.ActivityEnterLocalGameBinding
import com.google.android.gms.tasks.Task
import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.firestore.FieldPath
import com.google.firebase.firestore.FirebaseFirestore
import com.google.firebase.firestore.QuerySnapshot
import java.util.*

class EnterLocalGameActivity : AppCompatActivity() {

    private var firestoreDatabase: FirebaseFirestore? = null
    private lateinit var screen: ActivityEnterLocalGameBinding

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        screen = ActivityEnterLocalGameBinding.inflate(layoutInflater)
        setContentView(screen.root)
        initGUI()
        hideSystemUI()
    }

    /**
     * Método chamado quando a janela atual da activity ganha ou perde o foco, é utilizado para es-
     * conder novamente a barra de status e a navigation bar.
     * @param hasFocus - booleano que indica se a janela desta atividade tem foco.
     */
    override fun onWindowFocusChanged(hasFocus: Boolean) {
        super.onWindowFocusChanged(hasFocus)
        hideSystemUI()
    }

    /**
     * Método utilizado para fazer com que a barra de status e a navigation bar não sejam exibidas
     * na activity. Caso o usuário queira visualizá-las, ele deve realizar um movimento de arrastar
     * para cima (na navigation bar), ou para baixo (na status bar), o que fará com que elas apare-
     * çam por um momento e depois sumam novamente.
     */
    private fun hideSystemUI() {
        WindowCompat.setDecorFitsSystemWindows(window, false)
        val controller = WindowCompat.getInsetsController(window, window.decorView)
        if (controller != null) {
            controller.hide(WindowInsetsCompat.Type.statusBars())
            controller.hide(WindowInsetsCompat.Type.navigationBars())
            controller.systemBarsBehavior = WindowInsetsControllerCompat.BEHAVIOR_SHOW_TRANSIENT_BARS_BY_SWIPE
        }
    }

    fun initGUI() {
        firestoreDatabase = FirebaseFirestore.getInstance()
        screen.joinRoomBTN.setOnClickListener {
            if (screen.roomCodeEDT.editText?.text.toString().isEmpty()) {
                Toast.makeText(this@EnterLocalGameActivity, "Informe um código de sala.", Toast.LENGTH_LONG).show()
            } else {
                joinGameRoom(screen.roomCodeEDT.editText!!.text.toString())
            }
        }
    }
    @Suppress("UNCHECKED_CAST")
    private fun joinGameRoom(roomCode: String?) {
        firestoreDatabase!!.collection("partidaLocal").whereEqualTo(FieldPath.documentId(), roomCode).get()
            .addOnCompleteListener { task: Task<QuerySnapshot?> ->
                if (task.isSuccessful) {
                    if (task.result?.documents?.isNotEmpty() == true) {
                        val document = task.result?.documents!![0]
                        if (document.exists()) {
                            val qntdPlayers = document["qntd"].toString().toInt()
                            if (qntdPlayers >= 4) {
                                Toast.makeText(this@EnterLocalGameActivity, "Essa sala já está cheia!", Toast.LENGTH_LONG).show()
                            } else {
                                val nomeJogadores = (document["nomeJogadores"] as ArrayList<String?>?)!!
                                nomeJogadores[qntdPlayers] = FirebaseAuth.getInstance().currentUser?.displayName
                                firestoreDatabase!!.collection("partidaLocal").document(document.id).update("nomeJogadores", nomeJogadores)
                                val uidJogadores = (document["uidJogadores"] as ArrayList<String>?)!!
                                uidJogadores[qntdPlayers] = FirebaseAuth.getInstance().currentUser!!.uid
                                firestoreDatabase!!.collection("partidaLocal").document(document.id).update("uidJogadores", uidJogadores)
                                firestoreDatabase!!.collection("partidaLocal").document(document.id).update(
                                    "qntd", (Objects.requireNonNull(
                                        document["qntd"]
                                    ).toString().toInt() + 1).toString()
                                )
                                val troca = Intent(this@EnterLocalGameActivity, LocalGameActivity::class.java)
                                troca.putExtra("matchCreator", false)
                                troca.putExtra("roomCode", roomCode)
                                startActivityForResult(troca, 999)
                            }
                        }
                    } else {
                        Toast.makeText(this@EnterLocalGameActivity, "Essa sala não existe!", Toast.LENGTH_LONG).show()
                    }
                } else {
                    Log.d("Erro ao entrar na sala", "Erro: ", task.exception)
                }
            }
    }
}